File: DataStructure/merge-sort/1.py
def merge_sort(list):
    n = len(list)

    if n <= 1:
        return

    mid = n // 2
    g1 = list[:mid]
    g2 = list[mid:]

    merge_sort(g1)
    merge_sort(g2)

    i1 = 0
    i2 = 0
    ia = 0

    while i1 < len(g1) and i2 < len(g2):
        if g1[i1] < g2[i2]:
            list[ia] = g1[i1]
            i1 += 1
            ia += 1
        else:
            list[ia] = g2[i2]
            i2 += 1
            ia += 1

    while i1 < len(g1):
        list[ia] = g1[i1]
        i1 += 1
        ia += 1

    while i2 < len(g2):
        list[ia] = g2[i2]
        i2 += 1
        ia += 1


d = [6, 8, 3, 9, 10, 1, 2, 4, 7, 5]
merge_sort(d)
print(d)

File: data-structure/heap_sort/1.java
package DataStructure.heap_sort;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class solution {
    public static void main(String[] args) {
        List<Integer> heap_tree = new ArrayList<Integer>(Arrays.asList(new Integer[] { 5, 7, 1, 2, 6, 10 }));
        List<Integer> sorted = heap_sort(heap_tree);
        System.out.println(sorted);
    }

    public static List<Integer> heap_sort(List<Integer> tree) {
        List<Integer> sorted = new ArrayList<Integer>();
        int len = tree.size();
        while (len > 0) {
            build_heap(tree);
            swap(tree, 0, len - 1);
            sorted.add(0, tree.remove(len - 1));
            len--;
        }
        return sorted;
    }

    public static void build_heap(List<Integer> tree) {
        double mid = Math.floor(tree.size() / 2);
        while (mid >= 0) {
            System.out.println(mid);
            max_heapify(tree, (int) mid);
            mid--;
        }
    }

    public static void max_heapify(List<Integer> tree, int idx) {
        Integer node = idx;
        Integer left = idx * 2;
        Integer right = idx * 2 + 1;
        Integer len = tree.size() - 1;
        if (left > len || right > len) {
            return;
        } else {
            Integer bigger = tree.get(left) > tree.get(right) ? left : right;
            if (tree.get(bigger) < tree.get(node)) {
                return;
            }
            swap(tree, node, bigger);
            max_heapify(tree, bigger);
        }
    }

    public static void swap(List<Integer> array, Integer a, Integer b) {
        Integer temp = array.get(a);
        array.set(a, array.get(b));
        array.set(b, temp);
    }

}
